core/shared/src/main/scala/org/scalacheck/Prop.scala

```diff
@@ -784,7 +788,8 @@ object Prop {
     def shrinker(x: T, r: Result, shrinks: Int, orig: T): Result = {
       val xs = shrink(x)
       val res = r.addArg(Arg(labels,x,shrinks,orig,pp(x),pp(orig)))
-      if(xs.isEmpty) res else getFirstFailure(xs) match {
+      val originalException = Some(r.status).collect { case NonFatal(e) => e.getClass() }
```

Comment:
@armanbilge I think the culprit is here apparently.
Initially it was `case Exception` and I changed it to `NonFatal`. But I got confused by the name and  thought for a moment I was matching on a regular Exception (which does not really make sense in hindsight 😅), and just replaced it by `NonFatal`. Except it was actually 

`Exception` as in `Status` variant. So `originalException` is always empty because this never matches and `case (_, Result(Exception(e), _, _, _)) => !exceptionFilter.contains(e.getClass)` always return true. 

I'm not exactly sure why it hangs in the issue, but I imagine that if for some reason the stream of shrink is infinite (or long enough given some combination ?) then it will hang because as it is in case of an exception it just drops everything in the stream.

I cannot continue right now but I'll open a draft PR with the fix, and see if I can come up with more robust tests.

Reply:
@AurelienRichez thank you very much for looking into this so quickly! Please ping me when you put up your PR and I will take a closer look :)